File: Sources/WInstallerCore/LocalLogger.swift
import Foundation

/// A single local log entry. No ISO contents, license keys, or unrelated disk
/// contents are ever recorded (`TERMINAL_AUTOMATION.md` logging rules).
public struct LogEntry: Sendable, Codable, Equatable {
    public var timestamp: Date
    public var operationID: String
    public var tool: String
    public var arguments: [String]
    public var exitCode: Int32?
    public var userMessage: String
    public var technicalDetail: String?

    public init(
        timestamp: Date = Date(),
        operationID: String,
        tool: String,
        arguments: [String] = [],
        exitCode: Int32? = nil,
        userMessage: String,
        technicalDetail: String? = nil
    ) {
        self.timestamp = timestamp
        self.operationID = operationID
        self.tool = tool
        self.arguments = arguments
        self.exitCode = exitCode
        self.userMessage = userMessage
        self.technicalDetail = technicalDetail
    }
}

/// Local-only logging (`REQ-LOG-001..004`). Logs are stored under Application
/// Support and can be exported with home-directory paths redacted.
public actor LocalLogger {
    private let fileURL: URL?
    private var entries: [LogEntry] = []

    public init(directory: URL? = nil) {
        let base = directory ?? Self.defaultDirectory()
        if let base {
            try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
            self.fileURL = base.appendingPathComponent("winstaller.log")
        } else {
            self.fileURL = nil
        }
    }

    public func record(_ entry: LogEntry) {
        entries.append(entry)
        persist()
    }

    public var allEntries: [LogEntry] { entries }

    /// A redacted, human-readable transcript suitable for support bundles.
    public func exportText() -> String {
        entries.map(Self.format).joined(separator: "\n")
    }

    private func persist() {
        guard let fileURL else { return }
        let text = entries.map(Self.format).joined(separator: "\n")
        try? text.data(using: .utf8)?.write(to: fileURL, options: .atomic)
    }

    static func format(_ entry: LogEntry) -> String {
        let formatter = ISO8601DateFormatter()
        let time = formatter.string(from: entry.timestamp)
        let args = redact(entry.arguments.joined(separator: " "))
        let exit = entry.exitCode.map { " exit=\($0)" } ?? ""
        var line = "[\(time)] \(entry.operationID) \(entry.tool) \(args)\(exit) — \(entry.userMessage)"
        if let detail = entry.technicalDetail, !detail.isEmpty {
            line += "\n    " + redact(detail).replacingOccurrences(of: "\n", with: "\n    ")
        }
        return line
    }

    /// Redacts the current user's home directory and any `/Users/<name>` prefix.
    static func redact(_ string: String) -> String {
        var result = string
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        if !home.isEmpty {
            result = result.replacingOccurrences(of: home, with: "~")
        }
        if let regex = try? NSRegularExpression(pattern: "/Users/[^/ ]+", options: []) {
            let range = NSRange(result.startIndex..., in: result)
            result = regex.stringByReplacingMatches(in: result, options: [], range: range, withTemplate: "/Users/<redacted>")
        }
        return result
    }

    private static func defaultDirectory() -> URL? {
        guard let support = try? FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        ) else { return nil }
        return support.appendingPathComponent("wInstaller/logs", isDirectory: true)
    }
}

File: apps/windows/WInstaller.Core/LocalLogger.cs
using System.Text.RegularExpressions;

namespace WInstaller.Core;

/// <summary>
/// A single local log entry. No ISO contents, license keys, or unrelated disk
/// contents are ever recorded (TERMINAL_AUTOMATION.md logging rules).
/// </summary>
public sealed record LogEntry(
    string OperationId,
    string Tool,
    IReadOnlyList<string> Arguments,
    int? ExitCode,
    string UserMessage,
    string? TechnicalDetail = null)
{
    public DateTimeOffset Timestamp { get; init; } = DateTimeOffset.Now;
}

/// <summary>
/// Local-only logging (REQ-LOG-001..004). Logs are stored under Local AppData
/// and can be exported with user-profile paths redacted.
/// </summary>
public sealed class LocalLogger
{
    private readonly object _lock = new();
    private readonly string? _filePath;
    private readonly List<LogEntry> _entries = [];

    public LocalLogger(string? directory = null)
    {
        var baseDirectory = directory ?? DefaultDirectory();
        if (baseDirectory is not null)
        {
            try
            {
                Directory.CreateDirectory(baseDirectory);
                _filePath = Path.Combine(baseDirectory, "winstaller.log");
            }
            catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
            {
                _filePath = null;
            }
        }
    }

    public void Record(LogEntry entry)
    {
        lock (_lock)
        {
            _entries.Add(entry);
            Persist();
        }
    }

    public IReadOnlyList<LogEntry> AllEntries
    {
        get
        {
            lock (_lock)
            {
                return _entries.ToList();
            }
        }
    }

    /// <summary>A redacted, human-readable transcript suitable for support bundles.</summary>
    public string ExportText()
    {
        lock (_lock)
        {
            return string.Join("\n", _entries.Select(Format));
        }
    }

    private void Persist()
    {
        if (_filePath is null)
        {
            return;
        }
        try
        {
            File.WriteAllText(_filePath, string.Join("\n", _entries.Select(Format)));
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            // Logging must never take the pipeline down.
        }
    }

    public static string Format(LogEntry entry)
    {
        var time = entry.Timestamp.ToString("yyyy-MM-dd'T'HH:mm:ssK", System.Globalization.CultureInfo.InvariantCulture);
        var args = Redact(string.Join(" ", entry.Arguments));
        var exit = entry.ExitCode is { } code ? $" exit={code}" : "";
        var line = $"[{time}] {entry.OperationId} {entry.Tool} {args}{exit} — {entry.UserMessage}";
        if (!string.IsNullOrEmpty(entry.TechnicalDetail))
        {
            line += "\n    " + Redact(entry.TechnicalDetail!).Replace("\n", "\n    ");
        }
        return line;
    }

    private static readonly Regex UsersPathPattern = new(
        @"(?i)([a-z]):[\\/]users[\\/][^\\/\s]+",
        RegexOptions.Compiled);

    /// <summary>Redacts the current user's profile directory and any <c>C:\Users\name</c> path.</summary>
    public static string Redact(string text)
    {
        var result = text;
        var profile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        if (!string.IsNullOrEmpty(profile))
        {
            result = result.Replace(profile, "~");
        }
        return UsersPathPattern.Replace(result, @"$1:\Users\<redacted>");
    }

    private static string? DefaultDirectory()
    {
        var appData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
        return string.IsNullOrEmpty(appData) ? null : Path.Combine(appData, "wInstaller", "logs");
    }
}
